File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:

        from collections import Counter

        need = Counter(t)
        window = {}

        required = len(need)
        formed = 0

        left = 0

        minLen = float('inf')
        start = 0

        for right in range(len(s)):

            c = s[right]

            window[c] = window.get(c, 0) + 1

            if c in need and window[c] == need[c]:
                formed += 1

            while formed == required:

                if right - left + 1 < minLen:
                    minLen = right - left + 1
                    start = left

                remove = s[left]

                window[remove] -= 1

                if remove in need and window[remove] < need[remove]:
                    formed -= 1

                left += 1

        return "" if minLen == float('inf') else s[start:start + minLen]

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {

        unordered_map<char, int> need;
        unordered_map<char, int> window;

        for (char c : t)
            need[c]++;

        int required = need.size();
        int formed = 0;

        int left = 0;

        int minLen = INT_MAX;
        int start = 0;

        for (int right = 0; right < s.length(); right++) {

            char c = s[right];
            window[c]++;

            if (need.count(c) && window[c] == need[c])
                formed++;

            while (formed == required) {

                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    start = left;
                }

                char remove = s[left];

                window[remove]--;

                if (need.count(remove) &&
                    window[remove] < need[remove])
                    formed--;

                left++;
            }
        }

        return minLen == INT_MAX
               ? ""
               : s.substr(start, minLen);
    }
};